### PR TITLE
ldap_attrs: search_s based _is_value_present

### PR DESCRIPTION
**Pull request by mrvanes**
##### SUMMARY
Fixes https://github.com/ansible-collections/community.general/issues/977

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
community.general.ldap_attrs

##### ADDITIONAL INFORMATION
Instead of comparing the dn's attribute values using compare_s, I suggest using search_s plus filter, because search is X-ORDERED agnostic.

### CODE REVIEW

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment). Thanks.

**Comment by mrvanes** [comment]
I wonder why the tests fail on importing ldap in ldap_attrs.py? What can be done?

**Comment by mrvanes** [comment]
Before merging: I see that ldap_entry.py only excepts `ldap.NO_SUCH_OBJECT` on search_s and so raises `ldap.INVALID_DN_SYNTAX` on malformed dn. Do we want to break on malformed dn's so that the user is informed of bad config or silently ignore and return `_is_present = false` like I implemented?

File: changelogs/fragments/5385-search_s-based-_is_value_present.yaml
Comment (felixfontein) [inline changelogs/fragments/5385-search_s-based-_is_value_present.yaml:2]:
```suggestion
  - ldap_attrs - ... please write something here that regular users can understand who are not familiar with the implementation of this module ... (https://github.com/ansible-collections/community.general/pull/5385).
```

**Comment by mrvanes** [comment]
CI is happy and there's nothing I want to change. How to proceed from here?

**Comment by felixfontein** [comment]
@mrvanes basically this is waiting for a review by the module maintainers.

File: changelogs/fragments/5385-search_s-based-_is_value_present.yaml
Comment (felixfontein) [inline changelogs/fragments/5385-search_s-based-_is_value_present.yaml:2]:
```suggestion
  - ldap_attrs - fixes issue 977 by ignoring the ``{x}`` prefix on attribute values (https://github.com/ansible-collections/community.general/issues/977, https://github.com/ansible-collections/community.general/pull/5385).
```

**Comment by felixfontein** [comment]
At least for my use-case of the module, this doesn't seem to break something.

File: changelogs/fragments/5385-search_s-based-_is_value_present.yaml
Comment (russoz) [inline changelogs/fragments/5385-search_s-based-_is_value_present.yaml:2]: We usually place the link for the original issue in the `(...)` by the end.
```suggestion
  - ldap_attrs - fix issue by ignoring the ``{x}`` prefix on attribute values (https://github.com/ansible-collections/community.general/issues/977, https://github.com/ansible-collections/community.general/pull/5385).
```
These fragments end up in changelogs like this:
https://github.com/ansible-collections/community.general/blob/stable-5/CHANGELOG.rst#v5-7-0

Comment (russoz) [reply changelogs/fragments/5385-search_s-based-_is_value_present.yaml:2]: Other than this, LGTM

Comment (felixfontein) [reply changelogs/fragments/5385-search_s-based-_is_value_present.yaml:2]: Instead of just saying "fix issue" it should at least contain 1-2 words on what the issue is. In this case, it's related to ordering, so maybe "fix ordering issue"?

File: plugins/modules/net_tools/ldap/ldap_attrs.py
Comment (russoz) [inline plugins/modules/net_tools/ldap/ldap_attrs.py:269]: The `bool` is redundant in this statement, as the result of the `==` condition is already a `bool`. If you want to keep the `(...)` around it do clarify, fine:
```suggestion
            is_present = (len(dns) == 1)
```

Comment (mrvanes) [reply plugins/modules/net_tools/ldap/ldap_attrs.py:269]: I know it is redundant, I just mimicked the original `bool(self.connection.compare_s())` that also bools a returnvalue that is already a bool?

Comment (russoz) [reply plugins/modules/net_tools/ldap/ldap_attrs.py]: I had not looked much into the unchanged code in the module. Yes, it seems the the `compare_s()` function does return a `bool` [1], which makes the  `bool()` around it quite silly, IMO. If I may suggest, remove it from there as well.

[1] https://www.python-ldap.org/en/python-ldap-3.4.3/reference/ldap.html#ldap.LDAPObject.compare_s

**Comment by mrvanes** [comment]
So, while we're nitpicking over bools, I'd appreciate an opinion on this remark as well: https://github.com/ansible-collections/community.general/pull/5385#issuecomment-1283662063

File: plugins/modules/net_tools/ldap/ldap_attrs.py
Comment (felixfontein) [reply plugins/modules/net_tools/ldap/ldap_attrs.py]: Maybe some older version did not; I would not touch that line unless someone really verified that *every single version* of the Python LDAP bindings actually returns a boolean.

**Comment by felixfontein** [comment]
> `ldap.INVALID_DN_SYNTAX` on malformed dn. Do we want to break on malformed dn's so that the user is informed of bad config or silently ignore and return `_is_present = False` like I implemented?

I think the module should fail on invalid DNs and not ignore them silently.

File: plugins/modules/net_tools/ldap/ldap_attrs.py
Comment (mrvanes) [reply plugins/modules/net_tools/ldap/ldap_attrs.py]: Please pay attention here, that line is _replaced_ by my code...

**Comment by mrvanes** [comment]
> > `ldap.INVALID_DN_SYNTAX` on malformed dn. Do we want to break on malformed dn's so that the user is informed of bad config or silently ignore and return `_is_present = False` like I implemented?
> 
> I think the module should fail on invalid DNs and not ignore them silently.

Thx, that helps and I actually agree but wasn't sure.

**Comment by russoz** [comment]
> Before merging: I see that ldap_entry.py only catches `ldap.NO_SUCH_OBJECT` on search_s and so raises `ldap.INVALID_DN_SYNTAX` on malformed dn. Do we want to break on malformed dn's so that the user is informed of bad config or silently ignore and return `_is_present = False` like I implemented?

I'm sorry, I had missed that one. I agree with @felixfontein and your position.

**Review by russoz — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
@mrvanes thanks for your contribution!
@russoz thanks for reviewing!

**Pull request merged**
